evalml/tests/pipeline_tests/test_pipelines.py

```diff
@@ -344,8 +344,8 @@ def test_make_pipeline_from_components(X_y_binary, logistic_regression_binary_pi
     est = RandomForestClassifier(random_state=7)
     pipeline = make_pipeline_from_components([imp, est], ProblemTypes.BINARY, custom_name='My Pipeline',
                                              random_state=15)
-    assert [c.__class__ for c in pipeline.component_graph] == [Imputer, RandomForestClassifier]
```

Comment:
I think this highlights a subtle breaking change. Before `component_graph` would be populated with instances of the components after init. Now `component_graph` will always be the "static" graph definition (no instances). Is there a reason the instances are kept private as `_component_graph`?

If there is a reason to keep the instances separate from the "static" graph, I'd prefer if the instances were kept public like before. In that case, we can rename `_component_graph` to `component_instances` or something else?

Reply:
This is a really interesting point. The challenge here is that the instances are contained within the `ComponentGraph` object, which we're keeping completely private from users. It may be possible to add a separate public reference to the instances directly from the pipeline, but I worry that between this and the `linearized_component_graph` classproperty we discussed earlier, we're adding a fair amount of redundant information that's only there to maintain the current status quo.

What do you think?

Reply:
I see what you mean about wanting to keep `ComponentGraph` private from the end user. I think you're right that we shouldn't add another accessor for the instances. Maybe what we should do is just make iterating over/accessing the pipeline instead of the pipeline.component_graph the "preferred" public accessor for the component instances. The benefit is that this functionality is already built into your PR!

For example, the line we're commenting on would turn to `[c.__class__ for c in pipeline]`. 

This would require updating our [docs](https://evalml.alteryx.com/en/stable/user_guide/pipelines.html) to not access instances with `pipeline.component_graph` and our unit tests as well. 

What do you think? If you are on board, we should leave this for another issue since this diff is large already.

Reply:
I like this idea, I think it makes a lot of sense! I do agree it's better fit for a separate issue though haha, so I can file that if other people also think this would be beneficial!

Reply:
Filed issue #1571